mesonbuild/dependencies/base.py

```diff
@@ -2137,6 +2140,79 @@ def log_tried(self):
+    :*_class: A *type* or callable that creates a class, and has the
+        signature of an ExternalDependency
+    :system_class: If you pass DependencyMethods.SYSTEM in methods, you must
+        set this argument.
+    """
+
+    def __init__(self, name: str, methods: T.List[DependencyMethods], *,
+                 pkgconfig_name: T.Optional[str] = None,
+                 pkgconfig_class: T.Type[PkgConfigDependency] = PkgConfigDependency,
+                 cmake_name: T.Optional[str] = None,
+                 cmake_class: T.Type[CMakeDependency] = CMakeDependency,
+                 configtool_class: T.Optional[T.Type[ConfigToolDependency]] = None,
+                 framework_name: T.Optional[str] = None,
+                 framework_class: T.Type[ExtraFrameworkDependency] = ExtraFrameworkDependency,
+                 system_class: T.Type[ExternalDependency] = ExternalDependency):
+
+        if DependencyMethods.CONFIG_TOOL in methods and not configtool_class:
+            raise DependencyException('A configtool must have a custom class')
+
+        self.methods = methods
+        self.classes = {
+            # Just attach the correct name right now, either the generic name
+            # or the method specific name.
+            DependencyMethods.EXTRAFRAMEWORK: functools.partial(framework_class, framework_name or name),
+            DependencyMethods.PKGCONFIG: functools.partial(pkgconfig_class, pkgconfig_name or name),
+            DependencyMethods.CMAKE: functools.partial(cmake_class, cmake_name or name),
+            DependencyMethods.SYSTEM: functools.partial(system_class, name),
+            DependencyMethods.CONFIG_TOOL: None,
+        }
+        if configtool_class is not None:
+            self.classes[DependencyMethods.CONFIG_TOOL] = functools.partial(configtool_class, name)
+
+    @staticmethod
+    def _process_method(method: DependencyMethods, env: Environment, for_machine: MachineChoice) -> bool:
+        """Report whether a method is valid or not.
+
+        If the method is valid, return true, otherwise return false. This is
+        used in a list comprehension to filter methods that are not possible.
+
+        By default this only remove EXTRAFRAMEWORK dependencies for non-mac platforms.
+        """
+        # Extra frameworks are only valid for macOS and other apple products
+        if (method is DependencyMethods.EXTRAFRAMEWORK and
+                not env.machines[for_machine].is_darwin()):
+            return False
+        return True
+
+    def __call__(self, env: Environment, for_machine: MachineChoice,
+                 kwargs: T.Dict[str, T.Any]) -> T.List['DependencyType']:
+        """Return a list of Dependencies with the arguments already attached."""
+        methods = process_method_kw(self.methods, kwargs)
+        possible = [m for m in self.methods if self._process_method(m, env, for_machine)]
+
+        return [functools.partial(self.classes[e], env, kwargs) for e in possible if e in methods]
```

Comment:
Writing this as a sequence of list comprehensions looks a bit odd. I think this is just testing if the method is selected (via the method: keyword, and possible), so could both could just be tested in the condition?)